near-sdk/src/environment/env.rs

```diff
@@ -1021,4 +1046,41 @@ mod tests {
             .build());
         assert_eq!(super::signer_account_pk(), key);
     }
+
+    #[test]
+    fn event_tests() {
+        use crate::test_utils::get_logs;
+
+        // Event without data
+        event("nepXXX", "1.0.0", "event_name", None, EventSerializationFormat::JSON);
+
+        // Event with string data
+        let data = "Swapped 1 token_in.near for 2 token_out.near";
+        event("nepXXX", "1.0.0", "swap", Some(data), EventSerializationFormat::JSON);
+
+        // Event with json data
+        let mut data = std::collections::HashMap::new();
+        data.insert("vector", vec![1, 2, 3, 4]);
+        event(
+            "nepXXX",
+            "1.0.0",
+            "json_data_event",
+            Some(&(serde_json::to_string(&data).unwrap())),
+            EventSerializationFormat::JSON,
+        );
```

Comment:
APIs like this are error-prone, and I would not teach our users bad practices.

Can you take inspiration from [FT](https://github.com/near/near-sdk-rs/blob/770cbce018a1b6c49d58276a075ace3da96d6dc1/near-contract-standards/src/fungible_token/events.rs) / [NFT](https://github.com/near/near-sdk-rs/blob/master/near-contract-standards/src/non_fungible_token/events.rs) events and design the API that would help people to design their events reliably? E.g. encourage people to define their events as enum, and bake it into the flow somehow; here is just an illustration of how that could look like:

```rust
enum Mintbase001Event {
    Mint { owner_id: AccountId, token_id: String },
    Burn { ... },
    Transfer { ... },
}

impl JsonEventEmit for Mintbase001Event { /* there will be a default `.emit()` implementation which just serializes `&self` and logs it */ }

impl Tokens {
    fn mint(&mut self, token: Token) {
         self.insert(token);
         Mintbase001Event::Mint { ... }.emit();
    }
}

self.tokens.mint(token);
```

Reply:
Sorry, I am new to open source dev. Is it best practice to `--amend` the commit since this was more of a proposal commit or to make a new commit?

Also, is a logging structure like this `Event::emit(MyEvent::Mint { ... })` acceptable, or is the structure above the only acceptable structure?

Reply:
@coleFD we will squash PR commits here, so feel free to just push new commits

Reply:
> Also, is a logging structure like this Event::emit(MyEvent::Mint { ... }) acceptable, or is the structure above the only acceptable structure?

That makes sense to me. My proposed structure was just an illustration of how to promote the events to be reliable instead of ad-hoc.